syntax = "proto2";

package tidl_meta_arch;

// Top level configuration for TIDL Meta Architechtures
message TIDLMetaArch {
  optional string name = 1;
  repeated TidlMaCaffeSsd caffe_ssd = 2;
  repeated TidlMaTfOdApiSsd tf_od_api_ssd = 3;
  repeated TidlMaSsd tidl_ssd = 4;
  repeated TidlMaFasterRcnn tidl_faster_rcnn = 5;
}

message TIDLNmsParam {
  // Threshold to be used in nms.
  optional float nms_threshold = 1 [default = 0.3];
  // Maximum number of results to be kept.
  optional int32 top_k = 2;
  // Parameter for adaptive nms.
  optional float eta = 3 [default = 1.0];
}

// Encode/decode type.
enum TIDLBoxCodeType {
CORNER = 1;
CENTER_SIZE = 2;
CORNER_SIZE = 3;
} 

enum TIDLScoreConverter {
IDENTITY = 0;
SIGMOID = 1;
SOFTMAX = 2;
} 


// Message that store parameters used by PriorBoxLayer
message PriorBoxParameter {

  // Minimum box size (in pixels). Required!
  repeated float min_size = 1;
  // Maximum box size (in pixels). Required!
  repeated float max_size = 2;
  // Various of aspect ratios. Duplicate ratios will be ignored.
  // If none is provided, we use default ratio 1.
  repeated float aspect_ratio = 3;
  // If true, will flip each aspect ratio.
  // For example, if there is aspect ratio "r",
  // we will generate aspect ratio "1.0/r" as well.
  optional bool flip = 4 [default = true];
  // If true, will clip the prior so that it is within [0, 1]
  optional bool clip = 5 [default = false];
  // Variance for adjusting the prior bboxes.
  repeated float variance = 6;
  // By default, we calculate img_height, img_width, step_x, step_y based on
  // bottom[0] (feat) and bottom[1] (img). Unless these values are explicitely
  // provided.
  // Explicitly provide the img_size.
  optional uint32 img_size = 7;
  // Either img_size or img_h/img_w should be specified; not both.
  optional uint32 img_h = 8;
  optional uint32 img_w = 9;

  // Explicitly provide the step size.
  optional float step = 10;
  // Either step or step_h/step_w should be specified; not both.
  optional float step_h = 11;
  optional float step_w = 12;

  // Offset to the top left corner of each cell.
  optional float offset = 13 [default = 0.5];
  optional uint32 num_keypoint = 14 [default = 0];
}

message TIDLOdSaveParams {
  // Output directory. If not empty, we will save the results.
  optional string output_directory = 1;
  // Output name prefix.
  optional string output_name_prefix = 2;
  // Output format.
  //    VOC - PASCAL VOC output format.
  //    COCO - MS COCO output format.
  optional string output_format = 3;
  // If you want to output results, must also provide the following two files.
  // Otherwise, we will ignore saving results.
  // label map file.
  optional string label_map_file = 4;
  // A file which contains a list of names and sizes with same order
  // of the input DB. The file is in the following format:
  //    name height width
  //    ...
  optional string name_size_file = 5;
  // Number of test images. It can be less than the lines specified in
  // name_size_file. For example, when we only want to evaluate on part
  // of the test images.
  optional uint32 num_test_image = 6;
  // The resize parameter used in saving the data.
  optional string resize_param = 7;
}

// Message that store parameters used by DetectionOutputLayer
message TIDLOdPostProc {
  // Number of classes to be predicted. Required!
  optional uint32 num_classes = 1;
  // If true, bounding box are shared among different classes.
  optional bool share_location = 2 [default = true];
  // Background label id. If there is no background class,
  // set it as -1.
  optional int32 background_label_id = 3 [default = 0];
  // Parameters used for non maximum suppression.
  optional TIDLNmsParam nms_param = 4;
  // Parameters used for saving detection results.
  optional TIDLOdSaveParams save_output_param = 5;
  // Type of coding method for bbox.
  optional TIDLBoxCodeType code_type = 6 [default = CORNER];
  // If true, variance is encoded in target; otherwise we need to adjust the
  // predicted offset accordingly.
  optional bool variance_encoded_in_target = 8 [default = false];
  // Number of total bboxes to be kept per image after nms step.
  // -1 means keeping all bboxes after nms step.
  optional int32 keep_top_k = 7 [default = -1];
  // Only consider detections whose confidences are larger than a threshold.
  // If not provided, consider all boxes.
  optional float confidence_threshold = 9; 
  // If true, visualize the detection results.
  optional bool visualize = 10 [default = false];
  // The threshold used to visualize the detection results.
  optional float visualize_threshold = 11;
  // If provided, save outputs to video file.
  optional string save_file = 12;
  // Total number of keypoints predicted along with each bounding box.
  optional uint32 num_keypoint = 13 [default = 0];
}

// Configuration for one  TIDL SSD HEAD
message TidlMaSsd {
  repeated string box_input = 1; // List of input tensors in the same order as prior_box_param
  repeated string class_input = 2; // List of input tensors in the same order as prior_box_param
  optional string output = 3; // the name of output tensor
  optional string name = 4; // the name of the head
  optional uint32 in_width = 5;
  optional uint32 in_height = 6;
  optional TIDLScoreConverter score_converter = 7 [default = SOFTMAX];
  repeated PriorBoxParameter prior_box_param = 8;
  optional TIDLOdPostProc detection_output_param = 9;
}


// Configuration for one  Caffe SSD HEAD
message TidlMaCaffeSsd {
  repeated string box_input = 1; // List of input tensors in the same order as prior_box_param
  repeated string class_input = 2; // List of input tensors in the same order as prior_box_param
  optional string output = 3; // the name of output tensor
  optional string name = 4; // the name of the head
  optional uint32 in_width = 5;
  optional uint32 in_height = 6;
  optional TIDLScoreConverter score_converter = 7 [default = SOFTMAX];
  repeated PriorBoxParameter prior_box_param = 8;
  optional TIDLOdPostProc detection_output_param = 9;
}


// Configuration for list of TF OD API SSD heades
message TidlMaTfOdApiSsd {
  optional string name = 1;
}

// Configuration for list of TIDL Faster RCNN heades
message TidlMaFasterRcnn {
  optional string name = 1;
}
